### PR TITLE
[SPARK-21318][SQL]Improve exception message thrown by `lookupFunction`

### PR DESCRIPTION
**Pull request by stanzhai**
## What changes were proposed in this pull request?

The function actually exists in current selected database, and it's failed to init during `lookupFunciton`, but the exception message is: 
```
This function is neither a registered temporary function nor a permanent function registered in the database 'default'.
```

This is not conducive to positioning problems. This PR fix the problem.

## How was this patch tested?

new test case + manual tests

### CODE REVIEW

**Comment by stanzhai** [comment]
cc @liancheng

**Comment by gatorsmile** [comment]
@stanzhai Could you please add test cases?

**Comment by gatorsmile** [comment]
ok to test

**Comment by kiszk** [comment]
Jenkins, retest this please

**Comment by stanzhai** [comment]
@gatorsmile 
Some test cases have been added.
Thanks for reviewing.

**Comment by gatorsmile** [comment]
ping @stanzhai

**Comment by stanzhai** [comment]
Hi @gatorsmile , I've added some test cases, and passed on my machine.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
The failures are related to the changes you made here. Could you fix them? Thanks!

**Comment by stanzhai** [comment]
fixed @gatorsmile . retest this please

**Comment by gatorsmile** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 1]
```diff
@@ -129,14 +129,14 @@ private[sql] class HiveSessionCatalog(
     Try(super.lookupFunction(funcName, children)) match {
       case Success(expr) => expr
       case Failure(error) =>
-        if (functionRegistry.functionExists(funcName)) {
-          // If the function actually exists in functionRegistry, it means that there is an
-          // error when we create the Expression using the given children.
+        if (super.functionExists(name)) {
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: This should be replaced by `functionExists(name)`

File: sql/hive/src/test/java/org/apache/spark/sql/hive/execution/UDAFEmpty.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDAFEvaluator;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+
+/**
+ * An empty UDAF that throws a semantic exception
+ */
+public class UDAFEmpty extends AbstractGenericUDAFResolver {
+    @Override
+    public GenericUDAFEvaluator getEvaluator(TypeInfo[] info) throws SemanticException {
+        throw new SemanticException("Can not get a evaluator of the empty UDAF");
```
Comment (gatorsmile) [inline sql/hive/src/test/java/org/apache/spark/sql/hive/execution/UDAFEmpty.java]: Nit: `a evaluator` -> `an evaluator`

**Comment by gatorsmile** [comment]
cc @stanzhai Could you address my above comments? Thanks!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala (hunk 1 above)
Comment (stanzhai) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: We should keep use `super.functionExists(name)`, we can not load  a Hive's built-in function  if replaced by `functionExists(name)` and `org.apache.spark.sql.AnalysisException: Undefined function: 'histogram_numeric'` will be thrown.

**Comment by stanzhai** [comment]
cc @gatorsmile Addressed. Review this please. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -1155,7 +1155,8 @@ class Analyzer(
     override def apply(plan: LogicalPlan): LogicalPlan = plan.transformAllExpressions {
       case f: UnresolvedFunction if !catalog.functionExists(f.name) =>
         withPosition(f) {
-          throw new NoSuchFunctionException(f.name.database.getOrElse("default"), f.name.funcName)
+          val db = f.name.database.getOrElse(catalog.getCurrentDatabase)
+          throw new NoSuchFunctionException(db, f.name.funcName)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: The issue has been resolved. Can you revert the changes?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 4]
```diff
@@ -129,14 +129,14 @@ private[sql] class HiveSessionCatalog(
...
           // We need to throw the original exception.
           throw error
         } else {
-          // This function is not in functionRegistry, let's try to load it as a Hive's
-          // built-in function.
+          // This function is not in functionRegistry or externalCatalog,
+          // let's try to load it as a Hive's built-in function.
           // Hive is case insensitive.
           val functionName = funcName.unquotedString.toLowerCase(Locale.ROOT)
           if (!hiveFunctions.contains(functionName)) {
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: We do not need to change the other parts. We just need to throw the exception in `failFunctionLookup(funcName)`, right?

Comment (stanzhai) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: Yes, that's right.

**Comment by stanzhai** [comment]
cc @gatorsmile changes in `sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala` has been reverted.

**Comment by gatorsmile** [comment]
@stanzhai Could we just fix it by changing `failFunctionLookup`?

**Comment by stanzhai** [comment]
It's not reasonable, `failFunctionLookup` throws `NoSuchFunctionException`.
The function actually exists in current selected database, we should throw the exception which is due to an initialization failure, but not `NoSuchFunctionException`.

**Comment by cloud-fan** [comment]
what's the status here?

**Comment by maropu** [comment]
@stanzhai ping

**Comment by stanzhai** [comment]
The issue has been addressed a long time ago @cloud-fan @maropu

**Comment by stanzhai** [comment]
@cloud-fan 

User's hive UDFs are registered in externalCatalog which not exists in functionRegistry.

It will throws a NoSuchFunctionException when an exception is encountered while loading a hive UDF.

But we should throw the original exception.

So, I just fix the issue by:

```
if (functionRegistry.functionExists(funcName)) {
  throw error
} else {
  ...
}
```

changed to:

```
if (super.functionExists(name)) {
  throw error
} else {
  ...
}
```

The following is implementation of `super.functionExists`

```
def functionExists(name: FunctionIdentifier): Boolean = {
  val db = formatDatabaseName(name.database.getOrElse(getCurrentDatabase))
  requireDbExists(db)
  functionRegistry.functionExists(name) ||
    externalCatalog.functionExists(db, name.funcName)
}
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalogSuite.scala [hunk 5]
```diff
@@ -1440,6 +1441,8 @@ abstract class SessionCatalogSuite extends AnalysisTest {
         }
 
         assert(cause.getMessage.contains("Undefined function: 'undefined_fn'"))
+        // SPARK-21318: the error message should contains the current database name
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalogSuite.scala]: what's the full error message?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/UDFSuite.scala [hunk 6]
```diff
@@ -193,4 +193,29 @@ class UDFSuite
       }
     }
   }
+
+  test("SPARK-21318: The correct exception message should be thrown " +
+    "if a UDF/UDAF has already been registered") {
+    val UDAFName = "empty"
+    val UDAFClassName = classOf[org.apache.spark.sql.hive.execution.UDAFEmpty].getCanonicalName
+
+    withTempDatabase { dbName =>
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/UDFSuite.scala]: why do we have to test it inside a database? can't the default database work?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalogSuite.scala (hunk 5 above)
Comment (stanzhai) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalogSuite.scala]: org.apache.spark.sql.AnalysisException: Undefined function: 'undefined_fn'. This function is neither a registered temporary function nor a permanent function registered in the database 'db1'.;

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/UDFSuite.scala (hunk 6 above)
Comment (stanzhai) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/UDFSuite.scala]: @cloud-fan I just copied and modified the code from another test case, the default database works well.

The test case has been simplified now.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 7]
```diff
@@ -131,14 +131,14 @@ private[sql] class HiveSessionCatalog(
     Try(super.lookupFunction(funcName, children)) match {
       case Success(expr) => expr
       case Failure(error) =>
-        if (functionRegistry.functionExists(funcName)) {
-          // If the function actually exists in functionRegistry, it means that there is an
-          // error when we create the Expression using the given children.
+        if (super.functionExists(name)) {
+          // If the function actually exists in functionRegistry or externalCatalog,
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: `If the function exists (either in functionRegistry or externalCatalog), ...`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 8]
```diff
@@ -131,14 +131,14 @@ private[sql] class HiveSessionCatalog(
...
+          // it means that there is an error when we create the Expression using the given children.
           // We need to throw the original exception.
           throw error
         } else {
-          // This function is not in functionRegistry, let's try to load it as a Hive's
-          // built-in function.
+          // This function is not in functionRegistry or externalCatalog,
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: `This function does not exist(neither in functionRegistry or externalCatalog) ...`

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master/2.4!

Thanks for your patience! This PR is open for over a year 😂
